File: stack/155.MinStack.go
package main

import (
	"fmt"
	"math"
)

type MinStack struct {
	stack []int
	min   int
}

func Constructor() MinStack {
	return MinStack{min: math.MaxInt32, stack: []int{}}
}

func (this *MinStack) Push(x int) {
	if this.min >= x {
		this.stack = append(this.stack, this.min)
		this.min = x
	}
	this.stack = append(this.stack, x)
}

func (this *MinStack) Pop() {
	_len := len(this.stack)
	last := this.stack[_len-1]
	this.stack = this.stack[:_len-1]
	if last == this.min {
		this.min = this.stack[len(this.stack)-1]
		this.stack = this.stack[:len(this.stack)-1]
	}
}

func (this *MinStack) Top() int {
	return this.stack[len(this.stack)-1]
}

func (this *MinStack) GetMin() int {
	return this.min
}

func main() {
	fmt.Println("s")
	m := Constructor()
	m.Push(-2)
	m.Push(0)
	m.Push(-3)
	print(m.GetMin())
	m.Pop()
	print(m.Top())
	print(m.GetMin())
}

File: stack/155.MinStack.py
import sys
from typing import List


class MinStack:
    def __init__(self) -> None:
        """
        initialize your data structure here.
        """
        self.stack: List[int] = []
        self.min = sys.maxsize

    def push(self, x: int) -> None:
        if self.min >= x:
            self.stack.append(self.min)
            self.min = x
        self.stack.append(x)

    def pop(self) -> int:
        t = self.stack.pop()
        if self.min == t:
            self.min = self.stack.pop()
        return t

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min


if __name__ == "__main__":
    m = MinStack()
    m.push(-2)
    m.push(0)
    m.push(-3)
    print(m.getMin())
    m.pop()
    print(m.top())
    print(m.getMin())
    # MinStack minStack = new MinStack();
    # minStack.push(-2);
    # minStack.push(0);
    # minStack.push(-3);
    # minStack.getMin();   --> Returns -3.
    # minStack.pop();
    # minStack.top();      --> Returns 0.
    # minStack.getMin();   --> Returns -2.
